<a href="https://colab.research.google.com/github/avinashkumar18/Lab/blob/main/MLT_Lab/Lab_6_Implementation_of_logistic_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[Github Link](https://github.com/avinashkumar18/Lab/blob/main/MLT_Lab/Lab_6_Implementation_of_logistic_regression.ipynb)

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# read dataset
url = "https://raw.githubusercontent.com/avinashkumar18/Lab/main/MLT_Lab/titanic_data.csv"
train = pd.read_csv(url)

In [3]:
train.head(20)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0             1         0       3  ...   7.2500   NaN         S
1             2         1       1  ...  71.2833   C85         C
2             3         1       3  ...   7.9250   NaN         S
3             4         1       1  ...  53.1000  C123         S
4             5         0       3  ...   8.0500   NaN         S
5             6         0       3  ...   8.4583   NaN         Q
6             7         0       1  ...  51.8625   E46         S
7             8         0       3  ...  21.0750   NaN         S
8             9         1       3  ...  11.1333   NaN         S
9            10         1       2  ...  30.0708   NaN         C
10           11         1       3  ...  16.7000    G6         S
11           12         1       1  ...  26.5500  C103         S
12           13         0       3  ...   8.0500   NaN         S
13           14         0       3  ...  31.2750   NaN         S
14           15         0       3  ...   7.8542   NaN         S
15           16         1       2  ...  16.0000   NaN         S
16           17         0       3  ...  29.1250   NaN         Q
17           18         1       2  ...  13.0000   NaN         S
18           19         0       3  ...  18.0000   NaN         S
19           20         1       3  ...   7.2250   NaN         C

[20 rows x 12 columns]

**Feature Engineering**

In [4]:
# fix predictor and response variables
df = train[['Survived','Pclass','Sex','Age','Fare']]

In [5]:
df.head(20)

Survived  Pclass     Sex   Age     Fare
0          0       3    male  22.0   7.2500
1          1       1  female  38.0  71.2833
2          1       3  female  26.0   7.9250
3          1       1  female  35.0  53.1000
4          0       3    male  35.0   8.0500
5          0       3    male   NaN   8.4583
6          0       1    male  54.0  51.8625
7          0       3    male   2.0  21.0750
8          1       3  female  27.0  11.1333
9          1       2  female  14.0  30.0708
10         1       3  female   4.0  16.7000
11         1       1  female  58.0  26.5500
12         0       3    male  20.0   8.0500
13         0       3    male  39.0  31.2750
14         0       3  female  14.0   7.8542
15         1       2  female  55.0  16.0000
16         0       3    male   2.0  29.1250
17         1       2    male   NaN  13.0000
18         0       3  female  31.0  18.0000
19         1       3  female   NaN   7.2250

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   Fare      891 non-null    float64
dtypes: float64(2), int64(2), object(1)
memory usage: 34.9+ KB


In [7]:
# Encoding - change male to 1 and female to 0
df['Sex'] = df['Sex'].apply(lambda Sex:1 if Sex=='male' else 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [8]:
# Data Imputation -Handling missing values
df['Age'].isnull().values.any()
df['Age'].isnull().values.sum()

177

In [9]:
df.isnull().values.any()
df.isnull().values.sum()

177

In [10]:
df['Age'] = df['Age'].fillna(df['Age'].median()) # outliers
df['Age'].isnull().values.sum()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0

In [11]:
df['Age'].info()

AttributeError: ignored

In [12]:
df['Age'].head(6)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    28.0
Name: Age, dtype: float64

In [13]:
df.isna().sum()

Survived    0
Pclass      0
Sex         0
Age         0
Fare        0
dtype: int64

In [14]:
# setting predictor(X) and Response(Y) variables
X = df.drop('Survived',axis=1)
Y = df['Survived']

In [15]:
# split X and Y
from sklearn .model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3, random_state = 25)

In [16]:
# calling Logistic Regression model
from sklearn.linear_model import LogisticRegression
logit =LogisticRegression()
logit.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [17]:
# computing predictions or Y hat 
Y_pred = logit.predict(X_test)
print(Y_pred)

[0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 0 0 0 1 0 1 0 1 1 1 1 1 0
 1 1 0 1 0 1 0 0 1 0 1 0 0 1 0 1 0 1 1 1 0 0 0 1 1 0 0 0 1 0 1 0 0 0 1 1 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 1 1 0 0 0 0 0 1 0 1 0 1 1 1 0 0 1 1 0 1 1
 0 1 0 0 0 1 1 1 0 0 1 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 1 1 1 0 0 0 1 0 1 1 0
 0 0 0 0 0 1 1 1 0 0 0 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0 0 0 1 0 1 0 0 0 0 0 0
 0 0 0 1 1 0 1 0 1 1 0 1 1 0 1 0 1 0 1 1 0 0 0 1 0 1 1 0 1 0 0 0 0 0 1 1 0
 0 1 0 0 0 0 0 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 1 0 0 0 1 1 0 0 0 0 1 1 0 0
 1 0 0 0 1 0 0 1 0]


In [18]:
# confusion martix
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(Y_test, Y_pred)
print(confusion_matrix)

[[136  29]
 [ 31  72]]


In [19]:
# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, Y_pred)

0.7761194029850746

In [20]:
# Classification Report
from sklearn.metrics import classification_report
report =classification_report(Y_test, Y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.81      0.82      0.82       165
           1       0.71      0.70      0.71       103

    accuracy                           0.78       268
   macro avg       0.76      0.76      0.76       268
weighted avg       0.78      0.78      0.78       268



In [21]:
'''
Implement Logistic Regression Mode
Accuracy Score of Logit
Implement Naive Bayes Regression
Accuracy
Finally write a note on which classifier gave a better accuracy.
'''

'\nImplement Logistic Regression Mode\nAccuracy Score of Logit\nImplement Naive Bayes Regression\nAccuracy\nFinally write a note on which classifier gave a better accuracy.\n'

In [22]:
train.head(6)

PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S
5            6         0       3  ...   8.4583   NaN         Q

[6 rows x 12 columns]

In [23]:
new_df = train[['Survived','Pclass','Sex','Age','Fare']]

In [24]:
new_df['Sex'] = new_df['Sex'].apply(lambda Sex:1 if Sex=='male' else 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [25]:
new_df.isna().sum()

Survived      0
Pclass        0
Sex           0
Age         177
Fare          0
dtype: int64

In [26]:
print('Mean   is: ',df['Age'].mean())
print('Median is: ',df['Age'].median())
print('Mode   is: ',df['Age'].mode())

Mean   is:  29.36158249158249
Median is:  28.0
Mode   is:  0    28.0
dtype: float64


# For the above dataset, both the median and mode values are equal

so, Instead of filling with mode, I tried to fill it with mean (which has different value)

In [27]:
# filling na with mode 
new_df['Age'] = new_df['Age'].fillna(new_df['Age'].mean())
new_df['Age'].head(6)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


0    22.000000
1    38.000000
2    26.000000
3    35.000000
4    35.000000
5    29.699118
Name: Age, dtype: float64

In [28]:
new_df['Age']=round(new_df['Age'],2)
new_df['Age'].head(6)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    29.7
Name: Age, dtype: float64

In [29]:
# Set X and Y for new dataframe
new_X = new_df.drop('Survived',axis=1)
new_Y = new_df['Survived']

In [30]:
# splitting new values of X and Y (age is filled with mode value)
new_X_train, new_X_test, new_Y_train, new_Y_test = train_test_split(new_X, new_Y, test_size=0.3, random_state = 25)

In [31]:
# Logistic Regression model for new X and Y
logit.fit(new_X_train, new_Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [32]:
# prediction of y hat
new_Y_pred = logit.predict(new_X_test)

In [33]:
# new confusion matrix
from sklearn.metrics import confusion_matrix
new_confusion_matrix = confusion_matrix(new_Y_test, new_Y_pred)
print(new_confusion_matrix)

[[136  29]
 [ 31  72]]


In [34]:
# Accuracy Score
from sklearn.metrics import accuracy_score
accuracy_score(Y_test, Y_pred)

0.7761194029850746

In [35]:
# Classification Report
from sklearn.metrics import classification_report
report =classification_report(Y_test, Y_pred)
print(report)

              precision    recall  f1-score   support

           0       0.81      0.82      0.82       165
           1       0.71      0.70      0.71       103

    accuracy                           0.78       268
   macro avg       0.76      0.76      0.76       268
weighted avg       0.78      0.78      0.78       268



filling Both median and mean gave same output and accuracies. 
Note: is this error? or filling na with median and mode was independent of accuracy for this dataset?

In [36]:
# importing naive bayes classifier
from sklearn.naive_bayes import GaussianNB
naive_bayes = GaussianNB()

In [37]:
# fitting into the model
naive_bayes.fit(X_train, Y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [38]:
#predictions for Y hat 
naive_Y_pred = naive_bayes.predict(X_test)
print(naive_Y_pred)

[0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 0 1 1 1 1 1 0 0 0 0 0 1 0 1 0 1 1 1 1 1 0
 1 1 0 1 0 1 0 0 1 0 1 0 0 1 0 1 0 1 1 1 0 1 0 1 1 0 0 0 1 0 1 0 0 0 1 1 0
 0 0 0 0 0 0 0 0 0 1 0 0 0 1 0 1 1 1 0 0 0 0 0 1 0 1 0 1 1 1 0 0 1 1 1 1 1
 0 1 0 0 0 1 1 1 1 0 1 0 0 1 0 0 0 0 0 1 0 0 0 1 0 0 1 1 1 1 0 0 1 0 1 1 0
 0 0 0 0 0 1 1 0 0 0 0 0 1 0 0 0 1 0 0 0 0 1 1 0 0 0 0 1 1 0 1 0 1 0 0 0 0
 0 0 0 1 1 0 1 0 1 1 0 1 1 1 1 1 1 0 1 1 0 0 0 1 0 1 1 0 1 0 0 0 0 0 1 1 0
 0 1 0 0 1 0 0 1 0 1 0 1 0 1 1 1 0 0 1 0 1 0 0 1 0 0 0 1 1 0 0 0 0 1 1 0 0
 1 0 0 0 1 0 0 1 1]


In [39]:
# confusion matrix
naive_confusion_matrix = confusion_matrix(Y_test, naive_Y_pred)
print(naive_confusion_matrix)

[[130  35]
 [ 27  76]]


Error faced: when every tie calling the confusion_matrix or any function for the new variable, we need to import 

In [40]:
# confusion matrix
from sklearn.metrics import confusion_matrix
naive_confusion_matrix = confusion_matrix(Y_test, naive_Y_pred)
print(naive_confusion_matrix)

[[130  35]
 [ 27  76]]


In [41]:
# Accuracy Score
from sklearn.metrics import accuracy_score
naive_accuracy_score = accuracy_score(Y_test, naive_Y_pred)
print(naive_accuracy_score)

0.7686567164179104


In [42]:
# Classification Report
from sklearn.metrics import classification_report
naive_report =classification_report(Y_test, naive_Y_pred)
print(naive_report)

              precision    recall  f1-score   support

           0       0.83      0.79      0.81       165
           1       0.68      0.74      0.71       103

    accuracy                           0.77       268
   macro avg       0.76      0.76      0.76       268
weighted avg       0.77      0.77      0.77       268



From the above program, for the given dataset, the accuracies are

**Logistic Regression = 0.78**

**Naive Bayes Classifier = 0.77**

Both gave almost overall same accuracy with very small difference